### ISSUE TITLE
how to use get_trace_graph to get the attribute of a node in trace graph in pytorch 1.6?

### ISSUE THREAD

**Issue by ioou** [pytorch/pytorch#42505]
In pytorch 1.1 when using get_trace_graph to trace the model, we could get the node's attibution as follows:
%469 : bool = prim::Constant[value=0](), scope: Model/Backbone[backbone]/ConvBn[conv1]/BatchNorm2d[bn]
Then we could get the the attribution of the Node: backbone.conv1.bn and its class BatchNorm.

Then how to get the attribution name of the node in pytorch 1.6, as we can only get
%23 : float = prim::Constant[value=1]() # test.py:12:23

so how to  get the attribution in pytorch 1.6

cc @suo @gmagogsfm

**Comment by agolynski** [comment]
@ioou Could you give more context for this issue?

**Comment by ioou** [comment]
I want to trace the model to get the reference of the input layer of a given layer in the model.
if given a model, 
```
import torch
import torch.nn as nn
class Backbone(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 3, 3)
        self.bn1 = nn.BatchNorm2d(3)
    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu1(x)
        return x
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = Backbone()
    def forward(self, x):
        x = self.backbone(x)
        return x
model = Model()
im = torch.ones(1, 3, 28, 28)        
traced_net = torch.jit.get_trace_graph(model, im)
print(traced_net[0].graph())
```
In pytorch1.1,
```
The result is:
graph(%0 : Float(1, 3, 28, 28),
      %1 : Float(3, 3, 3, 3),
      %2 : Float(3),
      %3 : Float(3),
      %4 : Float(3),
      %5 : Float(3),
      %6 : Float(3),
      %7 : Long()):
  %8  : Float(1, 3, 28, 28) = aten::clone(%0)
  %9 : Float(3, 3, 3, 3) = aten::clone(%1)
  %10 : Float(3) = aten::clone(%2)
  %11 : Float(3) = aten::clone(%3)
  %12 : Float(3) = aten::clone(%4)
  %13 : Float(3) = aten::clone(%5)
  %14 : Float(3) = aten::clone(%6)
  %15 : Long() = aten::clone(%7)
  %16 : int = prim::Constant[value=1](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %17 : int = prim::Constant[value=1](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %18 : int[] = prim::ListConstruct(%16, %17), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %19 : int = prim::Constant[value=0](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %20 : int = prim::Constant[value=0](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %21 : int[] = prim::ListConstruct(%19, %20), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %22 : int = prim::Constant[value=1](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %23 : int = prim::Constant[value=1](), scope: Model/Backbone[backbone]/Conv2d[conv1]
  %24 : int[] = prim::ListConstruct(%22, %23), scope: Model/Backbone[backbone]/Conv2d[conv1]
...
  %input.1 : Float(1, 3, 26, 26) = aten::_convolution(%0, %1, %2, %18, %21, %24, %25, %28, %29, %30, %31, %32), scope: Model/Backbone[backbone]/Conv2d[conv1]
...
  %input : Float(1, 3, 26, 26) = aten::batch_norm(%input.1, %3, %4, %5, %6, %53, %54, %55, %56), scope: Model/Backbone[backbone]/BatchNorm2d[bn1]
  %58 : Float(1, 3, 26, 26) = aten::relu(%input), scope: Model/Backbone[backbone]/ReLU[relu1]
  return (%58)
```
With the scope information, we could get the attribution of  node '%input.1' is model.backbone.conv1, and its input is %0. In line '%input',  we could get its reference "model.backbone.bn1" and its input "%input.1" with the reference model.backbone.conv1.
However in pytorch 1.6, we could not get the scope information, 
if we use `traced_net = torch.jit._get_trace_graph(model, im)`  instead of` traced_net = torch.jit.get_trace_graph(model, im)`
```
The result is, 
graph(%input.1 : Float(1:2352, 3:784, 28:28, 28:1),
      %1 : Float(3:27, 3:9, 3:3, 3:1),
      %2 : Float(3:1),
      %3 : Float(3:1),
      %4 : Float(3:1),
      %5 : Float(3:1),
     ...
  %24 : int = prim::Constant[value=1]() # /home/yuanshiwei/anaconda3/lib/python3.6/site-packages/torch/nn/modules/conv.py:416:0
  %25 : int = prim::Constant[value=1]() # /home/yuanshiwei/anaconda3/lib/python3.6/site-packages/torch/nn/modules/conv.py:416:0
...
  %input.2 : Float(1:2028, 3:676, 26:26, 26:1) = aten::_convolution(%input.1, %1, %2, %26, %29, %32, %33, %36, %37, %38, %39, %40) # /home/yuanshiwei/anaconda3/lib/python3.6/site-packages/torch/nn/modules/conv.py:416:0
  %61 : bool = prim::Constant[value=1]()  # /home/yuanshiwei/anaconda3/lib/python3.6/site-packages/torch/nn/functional.py:2016:0
...
```
no explict information about scope. So how to get the name and its reference of each node of the traced graph in pytorch1.6?

**Comment by gmagogsfm** [comment]
Hi Ioou,

I don't think I fully understand your question, so please bear with me if this is incorrect. Sounds like you wanted to "callstack" of a node, ie which module/submodule it is created out of. 

I checked IR printer and logic to print `scope` still exists, just that most of the time `Graph` is created without specified `scope` like [here](https://github.com/pytorch/pytorch/blob/576aab5084268706d60820e6b1bc3056156e2589/torch/csrc/jit/frontend/ir_emitter.cpp#L3788).

We will look more into it.

Meanwhile, we would love to hear what is your specific use case for scope information, maybe we can propose a better solution?

**Comment by ioou** [comment]
hi, 

@gmagogsfm

1. The scope information is used to get the name of the node, so we can put the traced information(input and ouput layer name) to each module in the given model. So hope to get the name information of each node in the traced graph.

2. I want to get the previous and next node for each leaf module according to the trace.
For example as follows, for `model.backbone.bn1`, its prev and next nodes are `[model.backbone.conv1]` and` [model.backbone.relu1]` correspondingly. Node `model.backbone.concat `'s prev is ` [model.backbone.bn1, model.backbone.relu1]
`
```
class Concat(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x):  # x is a list of tensors
        return torch.cat(x, 1)
class Backbone(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 3, 3, padding=1)
        self.bn1 = nn.BatchNorm2d(3)
        self.relu1 = nn.ReLU()
        self.concat = Concat()
    def forward(self, x):
        x = self.conv1(x)
        x_bn = x = self.bn1(x)
        x = self.relu1(x)
        x = self.cat([x, x_bn])
        return x
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = Backbone()
    def forward(self, x):
        x = self.backbone(x)
        return x
```

**Comment by gmagogsfm** [comment]
Hi ioou,

Scope information isn't exactly part of the TorchScript IR, and it is difficult for us to provide backward compatibility guarantee or other types of support for it. 

Thanks for the clarification. I guess I am still not entirely sure about your bigger question. I understand that you wish to have node name info available, but for what bigger purpose? Maybe that's a question we could help answer.

**Comment by ioou** [comment]
hi, @gmagogsfm

I'd like to know the input layer and output layer of a given leaf layer.

Several relevant tasks depend on the graph of network. One of the them is to fuse the batch norm layer into it previous parameterized layer(conv,  linear). I have to know the previous layer of the bn to fuse bn's running mean, variance, w, b into the previous layer's parameter. Other tasks include transferring parameter info between layers according to their forward order.

**Comment by ioou** [comment]
hi,  @gmagogsfm 

I have found that name of node can be obtained by torch.jit.trace(model, tensor).graph_for(tensor). 

Thank you for your kind help.
